src/hardware/mouse/mouse_manymouse.cpp

```diff
@@ -0,0 +1,481 @@
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "mouse_manymouse.h"
+#include "mouse_common.h"
+#include "mouse_config.h"
+
+#include "callback.h"
+#include "checks.h"
+#include "math_utils.h"
+#include "pic.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+void manymouse_tick(uint32_t)
+{
+#if C_MANYMOUSE
+	ManyMouseGlue::GetInstance().Tick();
+#endif // C_MANYMOUSE
+}
+
+MousePhysical::MousePhysical(const std::string &name) : name(name) {}
+
+bool MousePhysical::IsMapped() const
+{
+	return mapped_id != MouseInterfaceId::None;
+}
+
+bool MousePhysical::IsDisconnected() const
+{
+	return disconnected;
+}
+
+MouseInterfaceId MousePhysical::GetMappedInterfaceId() const
+{
+	return mapped_id;
+}
+
+const std::string &MousePhysical::GetName() const
+{
+	return name;
+}
+
+ManyMouseGlue &ManyMouseGlue::GetInstance()
+{
+	static ManyMouseGlue manymouse_glue;
+	return manymouse_glue;
+}
+
+#if C_MANYMOUSE
+
+ManyMouseGlue::~ManyMouseGlue()
+{
+	PIC_RemoveEvents(manymouse_tick);
+	ManyMouse_Quit();
+}
+
+void ManyMouseGlue::InitIfNeeded()
+{
+	if (initialized || malfunction)
+		return;
+
+	num_mice = ManyMouse_Init();
+	if (num_mice < 0) {
+		LOG_ERR("MOUSE: ManyMouse initialization failed");
+		ManyMouse_Quit();
+		malfunction = true;
+		return;
+	}
+	initialized = true;
+
+	if (num_mice >= max_mice) {
+		num_mice           = max_mice - 1;
+		static bool logged = false;
+		if (!logged) {
+			logged = true;
+			LOG_ERR("MOUSE: Up to %d simultaneously connected mice supported",
+			        max_mice);
+		}
+	}
+
+	const auto new_driver_name = std::string(ManyMouse_DriverName());
+	if (new_driver_name != driver_name) {
+		driver_name = new_driver_name;
+		LOG_INFO("MOUSE: ManyMouse driver '%s'", driver_name.c_str());
+	}
+
+	Rescan();
+}
+
+void ManyMouseGlue::ShutdownIfSafe()
+{
+	if (is_mapping_in_effect || config_api_counter)
+		return;
+
+	ShutdownForced();
+}
+
+void ManyMouseGlue::ShutdownForced()
+{
+	if (!initialized)
+		return;
+
+	PIC_RemoveEvents(manymouse_tick);
+	ManyMouse_Quit();
+	ClearPhysicalMice();
+	num_mice    = 0;
+	initialized = false;
+}
+
+void ManyMouseGlue::StartConfigAPI()
+{
+	// Config API object is being created
+	++config_api_counter;
+	assert(config_api_counter > 0);
+}
+
+void ManyMouseGlue::StopConfigAPI()
+{
+	// Config API object is being destroyed
+	assert(config_api_counter > 0);
+	config_api_counter--;
+	ShutdownIfSafe();
+	if (!config_api_counter)
+		rescan_blocked_config = false;
+}
+
+void ManyMouseGlue::ClearPhysicalMice()
+{
+	mouse_info.physical.clear();
+	physical_devices.clear();
+	rel_x.clear();
+	rel_y.clear();
+}
+
+void ManyMouseGlue::Rescan()
+{
+	if (config_api_counter)
+		// Do not allow another rescan until MouseConfigAPI
+		// stops being used, it would be unsafe due to possible
+		// changes of physical device list size/indices
+		rescan_blocked_config = true;
+
+	ClearPhysicalMice();
+
+	for (uint8_t idx = 0; idx < num_mice; idx++) {
```

Comment:
`num_mice` is initialized in `ManyMouseGlue::InitIfNeeded()`, which has a protection - so it shouldn't happen. For extra safety I have reworked the code so that `num_mice` is `uint8_t` now.